### DISCUSSION TITLE
How do I deobfuscate Crashlytics logs that look like this?

### DISCUSSION THREAD

**Discussion by davidvorona** [invertase/react-native-firebase#5314] [Q&A]
We have a react-native app using a few modules in this library. The issue we're having is with Crashlytics: while it catches and reports crashes as expected, for some reason JS exceptions on Android are reported in a format that cannot be deobfuscated by [commonly discussed means](https://stackoverflow.com/questions/62214336/react-native-firebase-crashlytics-deobfuscation).

An example of a crash report:

```
Fatal Exception: com.facebook.react.common.JavascriptException
TypeError: Cannot read property 'id' of undefined This error is located at: in A in o in h in c in A in RCTView in Unknown in y in h in A in Unknown in A in Unknown in Unknown in Unknown in RCTView in Unknown in RCTView in Unknown in b in RCTView in Unknown in n in RCTView in Unknown in RCTView in Unknown in AnimatedComponent(View) in RCTView in Unknown in AnimatedComponent(View) in PanGestureHandler in Unknown in te in Unknown in RCTView in Unknown in Unknown in Unknown in Unknown in V in Unknown in Unknown in Unknown in RCTView in Unknown in RCTView in Unknown in RCTView in Unknown in Unknown in RCTView in Unknown in s in Unknown in PanGestureHandler in Unknown in RCTView in Unknown in s in Unknown in RCTView in Unknown in D in Unknown in RCTView in Unknown in Unknown in RCTView in Unknown in Unknown in H in c in Unknown in RCTView in Unknown in Unknown in S in Unknown in Unknown in Unknown in Unknown in Unknown in RCTView in Unknown in RCTView in Unknown in b in RCTView in Unknown in n in RCTView in Unknown in RCTView in Unknown in AnimatedComponent(View) in RCTView in Unknown in AnimatedComponent(View) in PanGestureHandler in Unknown in te in Unknown in RCTView in Unknown in Unknown in Unknown in Unknown in Unknown in Unknown in Unknown in Unknown in RCTView in Unknown in RCTView in Unknown in RCTView in Unknown in Unknown in RCTView in Unknown in s in Unknown in PanGestureHandler in Unknown in RCTView in Unknown in s in Unknown in RCTView in Unknown in D in Unknown in RCTView i<truncated: 1480 chars>
```

The link above mentions that this is the "raw" text, and that running the "formatted" text through `stack-beautifier` with a source map of our code will generate a more readable stack trace. Frankly, I'm not sure what the person answering the SO question is referring to. I see no "formatted" version of this crash log in Crashlytics, and what I've pasted above certainly cannot be deobfuscated.

On the other hand, JS exceptions reported from iOS devices *do* have a format that can be deobfuscated, and look like this:

```
Fatal Exception: RCTFatalException: Unhandled JS Exception: RangeError: Out of memory
Unhandled JS Exception: RangeError: Out of memory, stack: E@1717:811 _@1717:716 p@90:497 callTimers@90:2600 value@39:3469 <unknown>@39:662 value@39:2459 value@39:634
```

Running the above log through `stack-beautifier` returns a very helpful trace of exactly where the crash happened in our app. 

So, what is up with JS exceptions on Android? I'm most confused as to why I don't see more questions about this particularly unhelpful error format online anywhere, it makes me think I'm missing something obvious. I would appreciate some advice on this problem, as we are finding crashes but can do nothing about them since the trace is entirely unhelpful.

**Comment by mikehardy** [comment]
First thought: hermes or no? The permutations possible on bundling the JS are nearly endless so it's hard to say exactly. The stack traces I got (or used to get - I'm using Hermes now!) were de-obfuscatable but I haven't looked in a while in my own project

**Reply by davidvorona** [reply]
We are using Hermes! Nothing else about our configuration strikes me as particularly unique, though I suppose that's relative. 

I imagine many React Native projects have switched to Hermes for the perceived performance benefits and all that, so it seems strange to me that I can't really find any other developers grappling with this issue.

**Reply by mikehardy** [reply]
Good question why it's not known more, I trawled through hermes repo and found this, worth investigation? https://github.com/facebook/hermes/issues/276

I have a dream (and have floated it past people that may be able to help) of adding a functions-hookable event to crashlytics so there could be a "firebase extension" where if you had your sourcemaps loaded in firebase storage with the right version names and there would be a cloud function hooked by crashlytics that auto-decoded them. No traction yet, but I'm interested in anything you find here in order to keep the info organized for eventual real documentation at minimum and maybe distributable solution in the best future state

**Reply by davidvorona** [reply]
Hm, if my logs looked like the ones in that Hermes issue (and the [symbolication](https://reactnative.dev/docs/next/symbolication) page they link to) then our problems would be solved. It's this strange format--no lines numbers or file names--that makes this so difficult.

Still, the fact that some developers *are* receiving deobfuscatable crash reports on the Hermes engine must confirm that the problem can't rest entirely on using Hermes (which I already suspected, seeing as this isn't a well-documented problem). There must be something else in our app causing this issue.

Honestly, looking at the full crash log, it seems to me the important bit (e.g. `stack: value@1451:1379...`) might be hidden behind the final bit of the log I originally posted: `...in RCTView i<truncated: 1480 chars>`. I think I'm answering my own question as I write this, but it seems that the Crashlytics limit on log length is what's making this so difficult. I realize this is a bit out of the scope of this project and this question, but you wouldn't happen to have insight into this?

**Reply by mikehardy** [reply]
I do not have more insight sorry - If you're inclined these modules are open

https://github.com/firebase/firebase-android-sdk/tree/master/firebase-crashlytics
https://github.com/firebase/firebase-ios-sdk/tree/master/Crashlytics

...you can put on the caving helmet and go spelunking perhaps. Could be confounded if the truncation happens on the receiving end in the cloud though, I'm not aware of a way to inspect what is happening there

**Reply by davidvorona** [reply]
[I went spelunking.](https://github.com/firebase/firebase-android-sdk/blob/4bc47dfed12be4d1a60ef6b83b1ea9e2341a6439/firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/log/QueueFileLogStore.java#L135)

My experience with native code roughly equates to my experience spelunking, but I get the sense these logs are truncated on the receiving end, in the cloud, and the truncation we're seeing here is meant to handle it more elegantly (even though the developer here admits to the opposite) before the log is written to the file and sent off. Either way, it seems this will be a complex issue to solve.

I will continue to look into this, and will update here if I find anything that could be helpful to others experiencing this.

**Reply by mikehardy** [reply]
Very interesting. And that's a great log message, glad the dev left that.

can you confirm those really were on one line? no line breaks? That doesn't seem right based on how we're logging these stacks:

https://github.com/invertase/react-native-firebase/blob/7117bf49ef90d7922b9c4360824508207b95e0c2/packages/crashlytics/android/src/main/java/io/invertase/firebase/crashlytics/ReactNativeFirebaseCrashlyticsModule.java#L175-L200

They should be formatted line by line and the stack isn't even that big ?
But the result is still truncation somehow. I'm not sure what we're doing wrong to end up with truncation 

Tangent questions:  What's with all the "Unknown"'s as well? If you enabled Hermes on iOS (it works, I use it) would iOS then also suffer this? You can try it maybe with a test crash button to force both unhandled crashes and native ones (our test app here has those right in the app.js for an example)

**Reply by davidvorona** [reply]
When I get a chance I will test Hermes on iOS and force some crashes. That would definitely be a helpful exercise.

**Reply by mikehardy** [reply]
I had the opportunity to ping a Crashlytics Googler at Google I/O for the features required to do an auto-deobfuscation Firebase Extension, for what it's worth. All that's missing is a functions event from crashlytics for crash events to make it technically possible
